cmd/binlog-collector/collector/collector.go

```diff
@@ -0,0 +1,222 @@
+	err := c.newDB()
+	if err != nil {
+		return errors.Wrap(err, "new db connection")
+	}
+	defer c.closeDB()
+
+	err = c.CollectBinLogs()
+	if err != nil {
+		return errors.Wrap(err, "collect binlog files:")
+	}
+
+	return nil
+}
+
+func (c *Collector) newDB() error {
+	host, err := c.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	pxc, err := db.NewPXC(host, c.pxcUser, c.pxcPass)
+	if err != nil {
+		return errors.Wrapf(err, "new manager with host %s", host)
+
+	}
+	c.db = pxc
+
+	return nil
+}
+
+func (c *Collector) closeDB() error {
+	return c.db.Close()
+}
+
+func (c *Collector) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+c.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	sort.Strings(nodes)
+	lastHost := ""
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ".")
+			lastHost = nodeArr[0]
+		}
+	}
+	if len(lastHost) == 0 {
+		return "", errors.New("cant find host")
+	}
+
+	return lastHost + "." + c.pxcServiceName, nil
+}
+
+func (c *Collector) CollectBinLogs() error {
+	// get last uploaded binlog file name
+	binlogName, err := c.db.GetBinLogName(c.lastSet)
+	if err != nil {
+		return errors.Wrap(err, "get binlog name by set")
+	}
+	list, err := c.db.GetBinLogList()
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+
+	upload := false
+	// if there are no uploaded files we going to upload every binlog file
+	if len(binlogName) == 0 {
+		upload = true
+	}
+
+	for _, binlog := range list {
+		if binlog == binlogName { // this check is for uploading starting from needed file
+			upload = true
+		}
+		if upload {
+			err = c.manageBinlog(binlog)
+			if err != nil {
+				return errors.Wrap(err, "manage binlog")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Collector) manageBinlog(binlog string) error {
+	set, err := c.db.GetGTIDSet(binlog)
+	if err != nil {
+		return errors.Wrap(err, "get GTID set")
+	}
+	var setBuffer bytes.Buffer
+	setBuffer.WriteString(set)
+
+	if setBuffer.Len() == 0 {
+		return nil
+	}
+
+	cmd := exec.Command("mysqlbinlog", "-R", "-h"+c.db.GetHost(), "-u"+c.pxcUser, "-p"+c.pxcPass, binlog)
+	out, err := cmd.StdoutPipe()
+	if err != nil {
+		return errors.Wrap(err, "get stdout pipe")
+	}
+
+	errOut, err := cmd.StderrPipe()
+	if err != nil {
+		return errors.Wrap(err, "get stderr pipe")
+	}
+	err = cmd.Start()
+	if err != nil {
+		return errors.Wrap(err, "run mysqlbinlog command")
+	}
+	defer cmd.Wait()
+
+	outBuf := bytes.NewBuffer(make([]byte, c.bufferSize))
+	wg := sync.WaitGroup{}
+	go func() {
+		wg.Add(1)
+		_, err = outBuf.ReadFrom(out)
+		if err != nil {
+			log.Println("ERROR: read to buffer:", err)
+		}
+		wg.Done()
+	}()
```

Comment:
1. If you want to use WaitGroup, use it properly. `wg.Add(1)` should be set before goroutine. See https://play.golang.org/p/krJtLX14E0p

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/622/commits/cd3616c628e436107e203cf61b418b4cfe69c1fb